vmagent got panic: runtime error: invalid memory address or nil pointer dereference

### Describe the bug

Our victoria-metrics-agent pod got CrashLoopBackOff because of `panic: runtime error: invalid memory address or nil pointer dereference` as below on one cluster. There was maxScrapeSize message before the panic.
From the panic log, seems something related to streamaggr handlling. We have streaming aggregation configured and the streaming aggregation is working good on this cluster while it was Running before hitting the CrashLoopBackOff, and working good on other clusters.

```
{"ts":"2025-03-10T01:20:46.354Z","level":"warn","caller":"VictoriaMetrics/lib/promscrape/scrapework.go:383","msg":"cannot scrape target \"http://1.3.19.13:8080/metrics\" ({instance=\"1.3.19.13:8080\",job=\"kube-state-metrics\"}) 1 out of 1 times during -promscrape.suppressScrapeErrorsDelay=0s; the last error: the response from \"http://1.3.19.13:8080/metrics\" exceeds -promscrape.maxScrapeSize or max_scrape_size in the scrape config (40000000 bytes). Possible solutions are: reduce the response size for the target, increase -promscrape.maxScrapeSize command-line flag, increase max_scrape_size value in scrape config for the given target"}
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x8 pc=0x835e2f]

goroutine 6 [running]:
github.com/VictoriaMetrics/VictoriaMetrics/lib/streamaggr.putRateAggrStateValue(...)
        github.com/VictoriaMetrics/VictoriaMetrics/lib/streamaggr/rate.go:32
github.com/VictoriaMetrics/VictoriaMetrics/lib/streamaggr.(*rateAggrSharedValue).reset(0xc00157ba70)
        github.com/VictoriaMetrics/VictoriaMetrics/lib/streamaggr/rate.go:74 +0x4f
github.com/VictoriaMetrics/VictoriaMetrics/lib/streamaggr.putRateAggrSharedValue(0xc00157ba70)
        github.com/VictoriaMetrics/VictoriaMetrics/lib/streamaggr/rate.go:11 +0x18
github.com/VictoriaMetrics/VictoriaMetrics/lib/streamaggr.(*rateAggrValue).flush(0xc00115f780, {0xcc9820?, 0xc0001a8968}, 0xc00b180ee0, {0xc0015d2a40, 0x9})
        github.com/VictoriaMetrics/VictoriaMetrics/lib/streamaggr/rate.go:126 +0x225
github.com/VictoriaMetrics/VictoriaMetrics/lib/streamaggr.(*aggrOutputs).flushState.func1({0xa64740, 0xc00115f790}, {0xa504a0?, 0xc002515a40})
        github.com/VictoriaMetrics/VictoriaMetrics/lib/streamaggr/output.go:115 +0x209
sync.(*Map).Range(0xc006974d50?, 0xc006974d50)
        sync/map.go:501 +0x1f8
github.com/VictoriaMetrics/VictoriaMetrics/lib/streamaggr.(*aggrOutputs).flushState(0xc000238fc0?, 0xc000298c80?)
        github.com/VictoriaMetrics/VictoriaMetrics/lib/streamaggr/output.go:94 +0x4e
github.com/VictoriaMetrics/VictoriaMetrics/lib/streamaggr.(*aggregator).flush(0xc000238fc0, 0xc21fb0, {0x128bf40?, 0x2e7ddb00?, 0x128bf40?}, 0xc0012548d0, 0x0)
        github.com/VictoriaMetrics/VictoriaMetrics/lib/streamaggr/streamaggr.go:935 +0x145
github.com/VictoriaMetrics/VictoriaMetrics/lib/streamaggr.(*aggregator).runFlusher(0x0?, 0xc21fb0, 0x0, 0x0, 0x0)
        github.com/VictoriaMetrics/VictoriaMetrics/lib/streamaggr/streamaggr.go:852 +0x587
github.com/VictoriaMetrics/VictoriaMetrics/lib/streamaggr.newAggregator.func3()
        github.com/VictoriaMetrics/VictoriaMetrics/lib/streamaggr/streamaggr.go:712 +0x2c
created by github.com/VictoriaMetrics/VictoriaMetrics/lib/streamaggr.newAggregator in goroutine 1
        github.com/VictoriaMetrics/VictoriaMetrics/lib/streamaggr/streamaggr.go:711 +0x26c5
```

### To Reproduce

described above. vmagent version: 1.112.0.

### Version

vmagent version: 1.112.0

### Logs

_No response_

### Screenshots

_No response_

### Used command-line flags

_No response_

### Additional information

_No response_

--- a/docs/changelog/CHANGELOG.md
+++ b/docs/changelog/CHANGELOG.md
@@ -18,6 +18,8 @@ See also [LTS releases](https://docs.victoriametrics.com/lts-releases/).
 
 ## tip
 
+* BUGFIX: [stream aggregation](https://docs.victoriametrics.com/stream-aggregation): fix panic on `rate` output. See [this issue](https://github.com/VictoriaMetrics/VictoriaMetrics/issues/8469).
+
 ## [v1.113.0](https://github.com/VictoriaMetrics/VictoriaMetrics/releases/tag/v1.113.0)
 
 Released at 2025-03-07

--- a/lib/streamaggr/avg.go
+++ b/lib/streamaggr/avg.go
@@ -10,7 +10,7 @@ func (av *avgAggrValue) pushSample(_ aggrConfig, sample *pushSample, _ string, _
 	av.count++
 }
 
-func (av *avgAggrValue) flush(_ aggrConfig, ctx *flushCtx, key string) {
+func (av *avgAggrValue) flush(_ aggrConfig, ctx *flushCtx, key string, _ bool) {
 	if av.count > 0 {
 		avg := av.sum / av.count
 		ctx.appendSeries(key, "avg", avg)

--- a/lib/streamaggr/count_samples.go
+++ b/lib/streamaggr/count_samples.go
@@ -8,7 +8,7 @@ func (av *countSamplesAggrValue) pushSample(_ aggrConfig, _ *pushSample, _ strin
 	av.count++
 }
 
-func (av *countSamplesAggrValue) flush(_ aggrConfig, ctx *flushCtx, key string) {
+func (av *countSamplesAggrValue) flush(_ aggrConfig, ctx *flushCtx, key string, _ bool) {
 	if av.count > 0 {
 		ctx.appendSeries(key, "count_samples", float64(av.count))
 		av.count = 0

--- a/lib/streamaggr/count_series.go
+++ b/lib/streamaggr/count_series.go
@@ -18,7 +18,7 @@ func (av *countSeriesAggrValue) pushSample(_ aggrConfig, _ *pushSample, key stri
 	}
 }
 
-func (av *countSeriesAggrValue) flush(_ aggrConfig, ctx *flushCtx, key string) {
+func (av *countSeriesAggrValue) flush(_ aggrConfig, ctx *flushCtx, key string, _ bool) {
 	if len(av.samples) > 0 {
 		ctx.appendSeries(key, "count_series", float64(len(av.samples)))
 		clear(av.samples)

--- a/lib/streamaggr/histogram_bucket.go
+++ b/lib/streamaggr/histogram_bucket.go
@@ -14,7 +14,7 @@ func (av *histogramBucketAggrValue) pushSample(_ aggrConfig, sample *pushSample,
 	av.h.Update(sample.value)
 }
 
-func (av *histogramBucketAggrValue) flush(c aggrConfig, ctx *flushCtx, key string) {
+func (av *histogramBucketAggrValue) flush(c aggrConfig, ctx *flushCtx, key string, _ bool) {
 	ac := c.(*histogramBucketAggrConfig)
 	shared := av.shared
 	if ac.useSharedState {

--- a/lib/streamaggr/last.go
+++ b/lib/streamaggr/last.go
@@ -12,7 +12,7 @@ func (av *lastAggrValue) pushSample(_ aggrConfig, sample *pushSample, _ string, 
 	}
 }
 
-func (av *lastAggrValue) flush(_ aggrConfig, ctx *flushCtx, key string) {
+func (av *lastAggrValue) flush(_ aggrConfig, ctx *flushCtx, key string, _ bool) {
 	if av.timestamp > 0 {
 		ctx.appendSeries(key, "last", av.last)
 		av.timestamp = 0

--- a/lib/streamaggr/max.go
+++ b/lib/streamaggr/max.go
@@ -14,7 +14,7 @@ func (av *maxAggrValue) pushSample(_ aggrConfig, sample *pushSample, _ string, _
 	}
 }
 
-func (av *maxAggrValue) flush(_ aggrConfig, ctx *flushCtx, key string) {
+func (av *maxAggrValue) flush(_ aggrConfig, ctx *flushCtx, key string, _ bool) {
 	if av.defined {
 		ctx.appendSeries(key, "max", av.max)
 		av.max = 0

--- a/lib/streamaggr/min.go
+++ b/lib/streamaggr/min.go
@@ -14,7 +14,7 @@ func (av *minAggrValue) pushSample(_ aggrConfig, sample *pushSample, _ string, _
 	}
 }
 
-func (av *minAggrValue) flush(_ aggrConfig, ctx *flushCtx, key string) {
+func (av *minAggrValue) flush(_ aggrConfig, ctx *flushCtx, key string, _ bool) {
 	if av.defined {
 		ctx.appendSeries(key, "min", av.min)
 		av.min = 0

--- a/lib/streamaggr/output.go
+++ b/lib/streamaggr/output.go
@@ -112,7 +112,7 @@ func (ao *aggrOutputs) flushState(ctx *flushCtx) {
 			outputs = av.blue
 		}
 		for i, o := range outputs {
-			o.flush(ao.configs[i], ctx, outputKey)
+			o.flush(ao.configs[i], ctx, outputKey, ctx.isLast)
 		}
 		av.mu.Unlock()
 		if ctx.isLast {
@@ -135,6 +135,6 @@ type aggrConfig interface {
 
 type aggrValue interface {
 	pushSample(aggrConfig, *pushSample, string, int64)
-	flush(aggrConfig, *flushCtx, string)
+	flush(aggrConfig, *flushCtx, string, bool)
 	state() any
 }

--- a/lib/streamaggr/quantiles.go
+++ b/lib/streamaggr/quantiles.go
@@ -18,7 +18,7 @@ func (av *quantilesAggrValue) pushSample(_ aggrConfig, sample *pushSample, _ str
 	av.h.Update(sample.value)
 }
 
-func (av *quantilesAggrValue) flush(c aggrConfig, ctx *flushCtx, key string) {
+func (av *quantilesAggrValue) flush(c aggrConfig, ctx *flushCtx, key string, _ bool) {
 	ac := c.(*quantilesAggrConfig)
 	if av.h != nil {
 		ac.quantiles = av.h.Quantiles(ac.quantiles[:0], ac.phis)

--- a/lib/streamaggr/rate.go
+++ b/lib/streamaggr/rate.go
@@ -71,9 +71,11 @@ func (v *rateAggrSharedValue) reset() {
 	v.deleteDeadline = 0
 	v.prevTimestamp = 0
 	putRateAggrStateValue(v.blue)
-	putRateAggrStateValue(v.green)
 	v.blue = nil
-	v.green = nil
+	if v.green != nil {
+		putRateAggrStateValue(v.green)
+		v.green = nil
+	}
 }
 
 type rateAggrStateValue struct {
@@ -114,7 +116,7 @@ func (av *rateAggrValue) pushSample(_ aggrConfig, sample *pushSample, key string
 	state.timestamp = sample.timestamp
 }
 
-func (av *rateAggrValue) flush(c aggrConfig, ctx *flushCtx, key string) {
+func (av *rateAggrValue) flush(c aggrConfig, ctx *flushCtx, key string, isLast bool) {
 	ac := c.(*rateAggrConfig)
 	var state *rateAggrStateValue
 	suffix := ac.getSuffix()
@@ -142,7 +144,12 @@ func (av *rateAggrValue) flush(c aggrConfig, ctx *flushCtx, key string) {
 		sv.prevTimestamp = state.timestamp
 		state.timestamp = 0
 		state.increase = 0
-		av.shared[sk] = sv
+		if isLast {
+			delete(av.shared, sk)
+			putRateAggrSharedValue(sv)
+		} else {
+			av.shared[sk] = sv
+		}
 	}
 
 	if countSeries == 0 {

--- a/lib/streamaggr/std.go
+++ b/lib/streamaggr/std.go
@@ -18,7 +18,7 @@ func (av *stdAggrValue) pushSample(_ aggrConfig, sample *pushSample, _ string, _
 	av.avg = avg
 }
 
-func (av *stdAggrValue) flush(c aggrConfig, ctx *flushCtx, key string) {
+func (av *stdAggrValue) flush(c aggrConfig, ctx *flushCtx, key string, _ bool) {
 	ac := c.(*stdAggrConfig)
 	if av.count > 0 {
 		suffix := ac.getSuffix()

--- a/lib/streamaggr/sum_samples.go
+++ b/lib/streamaggr/sum_samples.go
@@ -8,7 +8,7 @@ func (av *sumSamplesAggrValue) pushSample(_ aggrConfig, sample *pushSample, _ st
 	av.sum += sample.value
 }
 
-func (av *sumSamplesAggrValue) flush(_ aggrConfig, ctx *flushCtx, key string) {
+func (av *sumSamplesAggrValue) flush(_ aggrConfig, ctx *flushCtx, key string, _ bool) {
 	ctx.appendSeries(key, "sum_samples", av.sum)
 	av.sum = 0
 }

--- a/lib/streamaggr/total.go
+++ b/lib/streamaggr/total.go
@@ -47,7 +47,7 @@ func (av *totalAggrValue) pushSample(c aggrConfig, sample *pushSample, key strin
 	av.shared.lastValues[key] = lv
 }
 
-func (av *totalAggrValue) flush(c aggrConfig, ctx *flushCtx, key string) {
+func (av *totalAggrValue) flush(c aggrConfig, ctx *flushCtx, key string, isLast bool) {
 	ac := c.(*totalAggrConfig)
 	suffix := ac.getSuffix()
 	// check for stale entries
@@ -55,7 +55,7 @@ func (av *totalAggrValue) flush(c aggrConfig, ctx *flushCtx, key string) {
 	av.total = 0
 	lvs := av.shared.lastValues
 	for lk, lv := range lvs {
-		if ctx.flushTimestamp > lv.deleteDeadline {
+		if ctx.flushTimestamp > lv.deleteDeadline || isLast {
 			delete(lvs, lk)
 		}
 	}

--- a/lib/streamaggr/unique_samples.go
+++ b/lib/streamaggr/unique_samples.go
@@ -10,7 +10,7 @@ func (av *uniqueSamplesAggrValue) pushSample(_ aggrConfig, sample *pushSample, _
 	}
 }
 
-func (av *uniqueSamplesAggrValue) flush(_ aggrConfig, ctx *flushCtx, key string) {
+func (av *uniqueSamplesAggrValue) flush(_ aggrConfig, ctx *flushCtx, key string, _ bool) {
 	if len(av.samples) > 0 {
 		ctx.appendSeries(key, "unique_samples", float64(len(av.samples)))
 		clear(av.samples)

